proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+```
+
+`<k>` cell specifies that the execution eventually reaches RETURN instruction meaning that the program will successfully terminate. The RETURN instruction says that a 32-byte return value will be stored in the memory at the location RET_ADDR.
+
+```{.k .erc20-spec .hkg-spec}
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("totalSupply", .TypedArgs)
+```
+
+`<localMem>` cell specifies that the local memory is empty in the beginning, but in the end, it will contain some contents including the return value `TOTAL`, the total supply of the tokens.
+
+The other entries are represented by the wildcard symbol `_`, meaning that they are not relevant to the functional correctness of the program.
+
+```{.k .erc20-spec .hkg-spec}
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(TOTAL, 32) ]
+      _:Map
+    )
+```
+
+`<gas>` cell specifies the maximum gas amount. Here we give a loose upper-bound for the demonstration purpose. In practice, one should set a reasonable amount of the gas limit to see if the program does not consume too much gas (i.e., no gas leakage).
+
+```{.k .erc20-spec .hkg-spec}
+gas: {gascap} => _
+```
+
+`<log>` cell specifies that no log is generated during the execution.
+
+```{.k .erc20-spec .hkg-spec}
+log: _
+```
+
+`<refund>` cell specifies that no gas is refunded. Note that it does not mean it consumes all the provided gas. The gas refund is different from returning the remaining gas after the execution. It is another notion to capture some specific gas refund events that happen, for example, when an unused storage entry is re-claimed (i.e., garbage-collected). The following specification ensures that no such event happens during the execution of the current function.
+
+```{.k .erc20-spec .hkg-spec}
+refund: _
+```
+
+`<storage>` cell specifies that the value of `totaySupply` is `TOTAL` and other entries are not relevant (and could be arbitrary values). Specifying the irrelevant entries implicitly expresses the non-interference property. That is, the total supply value will be returned regardless of what the other entires of the storage are. This representation of the irrelevant part is used throughout the entire specification, ensuring one of the principal security properties.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(totalsupply)}, .IntList) |-> TOTAL
+    _:Map
+```
+
+The side-condition specifies the range of symbolic values based on their types.
+
+```{.k .erc20-spec .hkg-spec}
+requires:
+    andBool 0 <=Int TOTAL     andBool TOTAL     <Int (2 ^Int 256)
+
+comments:
+```
+
+
+
+
+balanceOf
+---------
+
+The specification of `balanceOf` is similar to that of `totalSupply`.
+
+```{.k .erc20-spec .hkg-spec}
+[balanceOf]
+
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("balanceOf", #address(OWNER))
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(BAL, 32) ]
```

Comment:
Not `[macro]` but `[function]`. But, yes, they are syntactic sugars.

Note that they are desugared into 32 map entries: `RET_ADDR |-> _ RET_ADDR + 1 |-> _` ... `RET_ADDR + 31 |-> _`. We do not want to write it directly because of two reasons: it is too verbose, and there is a performance bug in parsing them.

Reply:
OK, sounds good.  So the `[function]` rules are now applied statically?  Didn't know that.  What if they are recursive?  I think we may want to get back to `[macro]`s, as the `[function]` attribute started to be overused.  OK for now, just add a comment somewhere about this, so people who do ERC20 proofs are aware of these things.

Reply:
Um, sorry, I made you more confusing. I think I should say it as DSLs instead of syntactic sugars. The `[function]` rules are not applied statically (if you mean static being applied before starting the execution). So, no issue with the recursions. Indeed, they are defined recursively.